### DISCUSSION TITLE
@types/node "IncomingMessage" type should be refactored into union

### DISCUSSION THREAD

**Discussion by thw0rted** [DefinitelyTyped/DefinitelyTyped#54920] [Issues with a @types package]
Discussion for https://github.com/DefinitelyTyped/DefinitelyTyped/issues/15808

The short version is that

````ts
import * as http from 'http';

http.createServer((req, res) => {
  const url: string = req.url;
  console.log(url);
});
````

should compile and currently it does not, because `url` is optional, even for Server incoming messages where it cannot be undefined.  (Likewise for several other fields, more info in the source link.)

**Comment by aminnairi** [comment]
This has also the drawback of making the test coverage incomplete since there is no way to test that the `require("http").IncomingMessage.url` can be undefined or not.

One would want to silence the warning from TypeScript by writing something like the following code.

```typescript
import {createServer} from "http";
import {join} from "path";
import {URL} from "url";

const server = createServer((request, response) => {
  const url = new URL(join("http://localhost", request.url ?? ""));

  response.end(url.pathname);
});
```

Here, we use the null-coalescing operator to give a default value. But this leads to untested code due to the fact that this is impossible to reproduce from the POV of a client requesting our server.

One temporary solution might be to cast the property `url` to a string, this makes the code coverage aware that there is no conditionals and so nothing to test other than running the code.

```typescript
import {createServer} from "http";
import {join} from "path";
import {URL} from "url";

const server = createServer((request, response) => {
  const url = new URL(join("http://localhost", String(request.url)));

  response.end(url.pathname);
});
```

Notice here how we got from `request.url ?? ""` to `String(request.url)`.

Unless having an optional url can be reproduced, I doubt that the types from the `IncomingMessage` are well suited for when creating a server.
